gpkit/model.py

```diff
@@ -1,250 +1,424 @@
-        for var, sub in substitutions.items():
-            if is_sweepvar(sub):
-                del subs[var]
-                if isinstance(var, Strings):
-                    var = self.varkeys[var]
-                elif isinstance(var, Monomial):
-                    var = VarKey(var)
-                if isinstance(var, Iterable):
-                    suba = np.array(sub[1])
-                    if len(var) == suba.shape[0]:
-                        for i, v in enumerate(var):
-                            found_sweep.append(v)
-                            if hasattr(suba[i], "__call__"):
-                                self.linkedsweep.update({v: suba[i]})
-                            else:
-                                self.sweep.update({v: suba[i]})
-                    elif len(var) == suba.shape[1]:
-                        raise ValueError("whole-vector substitution"
-                                         "is not yet supported")
-                    else:
-                        raise ValueError("vector substitutions must share a"
-                                         "dimension with the variable vector")
-                elif hasattr(sub[1], "__call__"):
-                    found_sweep.append(var)
-                    self.linkedsweep.update({var: sub[1]})
-                else:
-                    found_sweep.append(var)
-                    self.sweep.update({var: sub[1]})
-
-        if not (subs or found_sweep):
-            raise KeyError("could not find anything"
-                           "to substitute in %s" % substitutions)
-
-        base = getattr(self, frombase)
-        substitutions = dict(base.substitutions)
-
-        if replace:
-            (varlocs,
-             exps,
-             cs) = self.unsubbed.varlocs, self.unsubbed.exps, self.unsubbed.cs
-            for sweepvar in found_sweep:
-                if sweepvar in substitutions:
-                    del substitutions[sweepvar]
-            if subs:
-                # resubstitute from the beginning, in reverse order
-                varlocs, exps, cs, subs = substitution(varlocs,
-                                                       self.varkeys,
-                                                       exps,
-                                                       cs,
-                                                       subs)
-                substitutions.update(subs)
-            try:
-                varlocs, exps, cs, subs = substitution(varlocs,
-                                                       self.varkeys,
-                                                       exps,
-                                                       cs,
-                                                       substitutions)
-            except KeyError:
-                # our new sub replaced the only previous sub
-                pass
+        self.substitutions.update(substitutions)
+        # edits in place...should it return a new model instead?
```

Comment:
In my opinion it's a tradeoff between:

Maintaining consistency with nomials (which are immutable, right?)

and

Speed, for certain applications (which we would need to profile)

Lacking the profile data I'm unsure what's best...